File: CPP/Offer/15.cpp
struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(NULL) {
	}
};

// 前向插入的思路
// 涉及双向操作的时候，注意避免链表的断裂
class Solution
{
  public:
    ListNode *ReverseList(ListNode *pHead)
    {
        if (pHead == nullptr || pHead->next == nullptr)
            return pHead;

        ListNode* pNode = pHead;       // 当前节点
        ListNode* pReHead = nullptr;   // 新链表的头结点

        while (pNode != nullptr)
        {
            ListNode* pNext = pNode->next;  // 保存下一个节点
            pNode->next = pReHead;
            pReHead = pNode;
            pNode = pNext;
        }
        return pReHead;
    }
};

File: Offer/Python/15.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    # 返回ListNode
    def ReverseList(self, pHead):
        if pHead == None or pHead.next == None:
            return pHead
        pNode, pReHead = pHead, None
        while pNode != None:
            pNext = pNode.next
            pNode.next = pReHead
            pReHead = pNode
            pNode = pNext
        return pReHead
